### PR TITLE
[McByte part 4] Add Cutie mask propagator

### PR DESCRIPTION
**Pull request by tstanczyk95**
## What does this PR do?

This PR adds a Cutie-based mask propagation component for McByte.

The new `CutieMaskPropagator` initializes Cutie memory from masks generated on a reference frame and propagates those masks to subsequent frames. The implementation integrates [Cutie](https://github.com/hkchengrex/Cutie) as an external dependency and follows the existing mask interfaces used by [McByte](https://github.com/tstanczyk95/McByte).

Main additions:

- Added CutieMaskPropagator
- Added automatic Cutie weight download with checksum verification
- Added utilities for:
    - mask format conversion between McByte and Cutie
    - tracklet-to-object ID mapping
    - mask confidence computation
    - Cutie package/config discovery
- Added validation checks for mask shapes and tracklet-mask mappings
- Added unit tests covering helper utilities and propagator behavior
- Added a visual integration script for manual validation of SAM initialization and Cutie mask propagation

## Type of Change

- New feature (non-breaking change that adds functionality)

## Testing

- [x] I have tested this change locally
- [x] I have added/updated tests for this change

**Test details:**
Added unit tests covering:

- RGB image to Cutie tensor conversion
- Cutie probability tensor conversion to indexed masks
- Conversion between indexed and binary mask formats
- Non-overlapping mask generation
- Tracklet-to-object mapping construction
- Mask confidence computation
- Propagator state reset behavior
- Input validation during initialization

Manual validation:

- Tested CutieMaskPropagator initialization on CPU
- Tested CutieMaskPropagator initialization on CUDA
- Verified automatic checkpoint download
- Verified Hydra configuration loading
- Verified end-to-end SAM to Cutie propagation using the visualization script on a multi-frame sequence

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors
- [ ] [McByte as a whole is still in the development phase] I have updated the documentation accordingly (if applicable)

## Additional Context

This PR is stacked on top of the SAM mask generator PR #441 and uses the masks produced by SAMBoxMaskGenerator for Cutie initialization.

It requires installation of torch, torch vision and SAM as specified in #441 (section Additional Context). 

Furthermore, Cutie must be installed in the following manner (adapted steps from the original repo):

- Go to your parent folder of `trackers`.
- `git clone https://github.com/hkchengrex/Cutie.git`
- `cd Cutie`
- Open `Cutie/pyproject.toml`. Remove: `'cchardet >= 2.1.7',` (whole line 43).
- `pip install -e .`

Tested on CPU and on GPU (CUDA 12.6) - separate Anaconda environments.

Visual test performed on [SoccerNet-tracking](https://www.soccer-net.org/tasks/tracking):
- sequence SNMOT-116
- frames from from 000548.jpg to 000750.jpg (inclusive)
- oracle detections converted to xyxy format

Example test run for the visual test (you can also run it just with a single bounding box):

```
python visual_tests/visualize_cutie_mask_propagator.py \
  --image-dir your_dataset_path/SoccerNet/tracking/test/SNMOT-116/img1 \
  --start-file 000548.jpg \
  --end-file 000750.jpg \
  --box 177,482,249,631 \
  --box 1241,569,1297,716 \
  --box 191,344,240,459 \
  --box 686,298,751,426 \
  --box 625,469,686,623 \
  --box 1382,378,1440,497 \
  --box 1220,480,1240,498 \
  --device cuda 
```

Analogously to `SAMBoxMaskGenerator` in #441, when `CutieMaskPropagator` is instantiated, it checks if the pre-trained Cutie weights file is already saved in the pointed location. If the model not saved, then it downloaded. Default location for the pre-trained model: `trackers_root/models/cutie/cutie-base-mega.pth`. Weighs from [here](https://github.com/hkchengrex/Cutie/blob/main/cutie/utils/download_models.py).

### CODE REVIEW

File: src/trackers/core/mcbyte/masks/cutie.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    "base-mega": "https://github.com/hkchengrex/Cutie/releases/download/v1.0/cutie-base-mega.pth",
+}
+
+CUTIE_CHECKPOINT_MD5S = {
+    "base-mega": "a6071de6136982e396851903ab4c083a",
+}
+
+CUTIE_DEFAULT_WEIGHTS_PATHS = {
+    "base-mega": Path("models/cutie/cutie-base-mega.pth"),
+}
```
Comment (Borda) [inline src/trackers/core/mcbyte/masks/cutie.py]: Let's use a similar pattern as supervsion handle assets

File: src/trackers/core/mcbyte/masks/cutie.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+    tensor format.
+    """
+    return (
+        torch.from_numpy(frame.transpose(2, 0, 1))
+        .float()
+        .to(
+            device,
+            non_blocking=True,
+        )
+        / 255.0
```
Comment (Borda) [inline src/trackers/core/mcbyte/masks/cutie.py]: I would rather check if the max value is above 1, to be sure we don't accidentally divide by 0-1 by 255

File: src/trackers/core/mcbyte/masks/cutie.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+            from cutie.inference.utils.args_utils import get_dataset_cfg
+            from cutie.model.cutie import CUTIE
+            from hydra import compose, initialize_config_dir
+            from omegaconf import open_dict
+        except ImportError as exc:
+            msg = (
+                "Cutie support requires Cutie to be installed. "
+                "Install Cutie separately, for example by cloning the Cutie "
+                "repository and running `pip install -e .` inside it."
+            )
```
Comment (Borda) [inline src/trackers/core/mcbyte/masks/cutie.py:239]: Let's add a todo note. This will be adjusted as these instructions are correlated with PR detection, not user will find it while running code

File: src/trackers/core/mcbyte/masks/cutie.py [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+        with initialize_config_dir(
+            version_base="1.3.2",
+            config_dir=str(self.config_path.resolve()),
+            job_name="eval_config",
+        ):
+            cfg = compose(config_name=config_name)
+
+        with open_dict(cfg):
+            cfg["weights"] = self.weights_path
```
Comment (Borda) [inline src/trackers/core/mcbyte/masks/cutie.py]: Maybe wrap this in a private method?

File: src/trackers/core/mcbyte/masks/cutie.py [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+        self,
+        frame: np.ndarray,
+    ) -> MaskOutput | None:
+        """Propagate initialized masks to the given frame."""
+        if not self._initialized:
+            return None
+
+        frame_torch = _image_to_torch(frame, self.device)
+
+        with torch.inference_mode():
```
Comment (Borda) [inline src/trackers/core/mcbyte/masks/cutie.py]: This could be decorator for whole method

File: src/trackers/core/mcbyte/masks/cutie.py [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+            tracklet_mask_dict=self._tracklet_object_dict.copy(),
+            mask_avg_prob_dict=mask_avg_prob_dict,
+        )
+
+    def _autocast_context(self) -> Any:
+        if self.use_amp:
+            if hasattr(torch, "amp"):
+                return torch.amp.autocast("cuda", enabled=True)
+            return torch.cuda.amp.autocast(enabled=True)
+        return nullcontext()
```
Comment (Borda) [inline src/trackers/core/mcbyte/masks/cutie.py]: I would make it plain func with `use_amp` arg

File: tests/core/test_mcbyte_cutie_mask_propagator.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        ),
+    )
+
+
+def test_build_tracklet_object_dict_converts_local_indices_to_cutie_object_ids() -> None:
+    tracklet_object_dict = _build_tracklet_object_dict(
+        {
+            10: 0,
+            20: 1,
+            30: 2,
```
Comment (Borda) [inline tests/core/test_mcbyte_cutie_mask_propagator.py]:
```suggestion
            30: 2
```
If you drop the last `,` linter will decide if needed or unwarp it... :)

File: tests/core/test_mcbyte_cutie_mask_propagator.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+        masks=np.zeros((2, 4, 4), dtype=bool),
+        tracklet_mask_dict={1: 1, 2: 2},
+        mask_avg_prob_dict=None,
+    )
+
+    with pytest.raises(ValueError, match="local mask indices"):
+        propagator.initialize(
+            frame=np.zeros((4, 4, 3), dtype=np.uint8),
+            mask_output=mask_output,
+        )
```
Comment (Borda) [inline tests/core/test_mcbyte_cutie_mask_propagator.py]: These seem very similar in logic, just different in/out, so let's use pytest.parametrize

File: src/trackers/core/mcbyte/masks/cutie.py [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+        mask_avg_prob_dict = _compute_mask_avg_prob_dict(
+            prob=prob,
+            object_ids=self._object_ids,
+        )
+
+        return MaskOutput(
+            masks=masks,
+            tracklet_mask_dict=self._tracklet_object_dict.copy(),
+            mask_avg_prob_dict=mask_avg_prob_dict,
+        )
```
[the remark this answers is not in the document]
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/cutie.py:372]: Good point. `CutieMaskPropagator` currently keeps Cutie object IDs internally, while MaskOutput is expected to expose local mask indices. Since the dynamic add/remove lifecycle in the next PR #459 changes this mapping more substantially and involves such behavior as ID changes/shifts, I will handle this consistently there rather than making a partial change here, in this PR. This will avoid introducing an intermediate mapping behavior that will be immediately replaced by the lifecycle implementation.

File: src/trackers/core/mcbyte/masks/cutie.py (hunk 1 above)
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/cutie.py]: Addressed in f8423b9.

File: src/trackers/core/mcbyte/masks/cutie.py (hunk 2 above)
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/cutie.py]: Addressed in f8423b9.

File: src/trackers/core/mcbyte/masks/cutie.py [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+        raise ValueError(f"Unsupported checkpoint URL scheme: {parsed_url.scheme!r}")
+
+    weights_path.parent.mkdir(parents=True, exist_ok=True)
+
+    logger.info("Downloading Cutie checkpoint to %s", weights_path)
+    _download_file(
+        checkpoint_url,
+        weights_path,
+        md5=CUTIE_CHECKPOINT_MD5S.get(model_type),
+    )
```
[the remark this answers is not in the document]
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/cutie.py:73]: Addressed in f8423b9.

File: tests/core/test_mcbyte_cutie_mask_propagator.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    indexed_mask = _torch_prob_to_numpy_mask(prob)
+
+    expected = np.array(
+        [
+            [0, 1],
+            [1, 2],
+        ],
+        dtype=np.uint8,
+    )
+    np.testing.assert_array_equal(indexed_mask, expected)
```
[the remark this answers is not in the document]
Comment (tstanczyk95) [reply tests/core/test_mcbyte_cutie_mask_propagator.py:63]: Addressed in f8423b9.

File: src/trackers/core/mcbyte/masks/cutie.py (hunk 5 above)
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/cutie.py]: Thanks for this suggestion!

Addressed in f8423b9.

Besides, it will be _also_ considered for `add_masks` and `remove_masks` in #459 (next PR).

File: src/trackers/core/mcbyte/masks/cutie.py (hunk 6 above)
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/cutie.py]: Addressed in f8423b9.

File: tests/core/test_mcbyte_cutie_mask_propagator.py (hunk 7 above)
Comment (tstanczyk95) [reply tests/core/test_mcbyte_cutie_mask_propagator.py]: Addressed in f8423b9.

File: tests/core/test_mcbyte_cutie_mask_propagator.py (hunk 8 above)
Comment (tstanczyk95) [reply tests/core/test_mcbyte_cutie_mask_propagator.py]: Addressed in f8423b9.

File: src/trackers/core/mcbyte/masks/cutie.py (hunk 4 above)
Comment (tstanczyk95) [reply src/trackers/core/mcbyte/masks/cutie.py]: Addressed in f8423b9.
